### PR TITLE
gRPC: Handle cancelled scope properly

### PR DESCRIPTION
**Pull request by oldergod**

### CODE REVIEW

File: wire-grpc-client/src/main/java/com/squareup/wire/GrpcMethod.kt [hunk 1]
```diff
@@ -116,16 +129,20 @@ internal sealed class GrpcMethod<S, R>(
...
     ): Any {
       val requestChannel = Channel<S>(0)
       val responseChannel = Channel<R>(0)
 
-      grpcClient.call(this, context, requestChannel, responseChannel)
+      val call = grpcClient.call(this, continuation, requestChannel, responseChannel)
 
-      return CoroutineScope(context).produce<Any> {
+      continuation.context[Job]!!.invokeOnCompletion { cause: Throwable? ->
+        if (cause is CancellationException) call.cancel()
```
Comment (oldergod) [inline wire-grpc-client/src/main/java/com/squareup/wire/GrpcMethod.kt]: TODO test still pass if that's commented out :/

File: build.gradle [hunk 2]
```diff
@@ -103,6 +103,10 @@ allprojects {
 
   repositories {
     mavenCentral()
+    // TODO(oldergod) remove once okHttp ships isDuplex
```
Comment (swankjesse) [inline build.gradle]: 👍🏻

File: wire-grpc-client/src/main/java/com/squareup/wire/GrpcMethod.kt [hunk 3]
```diff
@@ -94,17 +105,27 @@ internal sealed class GrpcMethod<S, R>(
...
+          call.cancel()
+          requestChannel.cancel()
+          responseChannel.cancel()
+        }
+      }
 
       return Pair(
           requestChannel,
-          CoroutineScope(context).async { responseChannel.consume { responseChannel.receive() } }
+          CoroutineScope(continuation.context).async {
```
Comment (swankjesse) [inline wire-grpc-client/src/main/java/com/squareup/wire/GrpcMethod.kt]: can you pull this out to be defined above? Seems awkward to have inline

File: wire-grpc-tests/src/test/java/com/squareup/wire/GrpcTest.kt [hunk 4]
```diff
@@ -145,4 +161,124 @@ class GrpcTest {
...
+
+  @Test
+  fun cancelStreamingRequest() {
+    mockService.enqueue(ReceiveCall("/routeguide.RouteGuide/RecordRoute"))
+    mockService.enqueue(Delay(500, TimeUnit.MILLISECONDS))
+    mockService.enqueue(ReceiveError)
+
+    runBlocking {
+      val deferred = async {
+        val result = routeGuideService.RecordRoute().second
```
Comment (swankjesse) [inline wire-grpc-tests/src/test/java/com/squareup/wire/GrpcTest.kt]: do you like this better?
```
val (_, receiveChannel) = routeGuideService.RecordRoute()
```

File: wire-grpc-tests/src/test/java/com/squareup/wire/GrpcTest.kt [hunk 5]
```diff
@@ -145,4 +161,124 @@ class GrpcTest {
...
+      }
+      delay(200)
+      deferred.cancel()
+      mockService.awaitSuccess()
+      assertThat(callReference.get()?.isCanceled).isTrue()
+    }
+  }
+
+  @Test
+  fun cancelChannel() {
```
Comment (swankjesse) [inline wire-grpc-tests/src/test/java/com/squareup/wire/GrpcTest.kt]: I wanna figure out a better way for this test to work. But I’m not sure how!

**Comment by swankjesse** [comment]
Looks like the new test is flaky:

```
> Task :wire-grpc-tests:test

com.squareup.wire.GrpcTest > cancelChannel FAILED
    java.lang.AssertionError at GrpcTest.kt:269

9 tests completed, 1 failed

> Task :wire-grpc-tests:test FAILED
```

**Pull request merged**